internal/rbd/rbd_util.go

```diff
@@ -800,18 +803,36 @@ func deleteSnapshot(ctx context.Context, vol *rbdVolume, pOpts *rbdSnapshot, cr
 	return err
 }
 
-// nolint: interfacer
-func restoreSnapshot(ctx context.Context, pVolOpts *rbdVolume, pSnapOpts *rbdSnapshot, cr *util.Credentials) error {
-	var output []byte
+func cloneRbdImageFromSnapshot(ctx context.Context, pVolOpts *rbdVolume, pSnapOpts *rbdSnapshot, cr *util.Credentials) error {
+	image := pVolOpts.RbdImageName
+	snapName := pSnapOpts.RbdSnapName
 
-	klog.V(4).Infof(util.Log(ctx, "rbd: clone %s using mon %s"), pVolOpts, pVolOpts.Monitors)
-	args := []string{"clone", pSnapOpts.String(), pVolOpts.String(),
-		"--id", cr.ID, "-m", pVolOpts.Monitors, "--keyfile=" + cr.KeyFile}
+	klog.V(4).Infof(util.Log(ctx, "rbd: clone %s/%s@%s %s using mon %s"), pVolOpts.Pool, pSnapOpts.RbdImageName, snapName, image, pVolOpts.Monitors)
 
-	output, err := execCommand("rbd", args)
+	options := librbd.NewRbdImageOptions()
+
+	if pVolOpts.ImageFeatures != "" {
+		features := imageFeaturesToUint64(ctx, pVolOpts.ImageFeatures)
+		err := options.SetUint64(librbd.RbdImageOptionFeatures, features)
+		if err != nil {
+			return errors.Wrapf(err, "failed to set image features")
+		}
+	}
+
+	err := options.SetUint64(imageOptionCloneFormat, 2)
+	if err != nil {
+		return errors.Wrapf(err, "failed to set image features")
+	}
+
+	ioctx, err := pVolOpts.conn.GetIoctx(pVolOpts.Pool)
```

Comment:
use `pVolOpts.openIoctx()` and `pVolOpts.ioctx` after that

Reply:
sorry already done, not pushed the latest change :(